### PR TITLE
Ignore tokens if their IDs in added token list are below the vocab's base size

### PR DESCRIPTION
**Pull request by seungduk-yanolja**
MistralOrca has added tokens that the IDs are below 30,000. It causes some issues when converting models.
https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/blob/main/added_tokens.json
Actually, these tokens are the same with the base model so they are unnecessary. But handling this kind of exception might be better.

### CODE REVIEW

**Comment by goerch** [comment]
This probably isn't the only problem. @staviq is working on similar [problems](https://github.com/ggerganov/llama.cpp/pull/3538). How do you validate the tokenizer of this model?

**Comment by staviq** [comment]
> Mistral Orca has added tokens that the ids are below 30,000. It causes some issues when converting models.

Can you provide link to the particular model ? I'd like to check this out

**Comment by seungduk-yanolja** [comment]
> > Mistral Orca has added tokens that the ids are below 30,000. It causes some issues when converting models.
> 
> Can you provide link to the particular model ? I'd like to check this out

https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/blob/main/added_tokens.json

**Comment by staviq** [comment]
> > > Mistral Orca has added tokens that the ids are below 30,000. It causes some issues when converting models.
> > 
> > 
> > Can you provide link to the particular model ? I'd like to check this out
> 
> https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/blob/main/added_tokens.json

That's the one I'm testing #3538 on, specifically because it has a mess in JSON files.

If you simply delete tokens 0,1,2 from added_tokens.json, everything works just fine, and those tokens are still in the vocab and are recognised. It makes no sense for them to even be there.

Not to mention they specify `<s>` `</s>` `<|im_start|>` `<|im_end|>` but then say bos is `<s>` but eos is `<|im_end|>` while still using `<|im_start|>`

**Comment by goerch** [comment]
> If you simply delete tokens 0,1,2 from added_tokens.json

According to the explanations in #3538 we'll have to ignore (or 'special';) case) them in the future, I think.

**Comment by seungduk-yanolja** [comment]
> > If you simply delete tokens 0,1,2 from added_tokens.json
> 
> According to the explanations in #3538 we'll have to ignore (or 'special';) case) them in the future, I think.

changed the code to ignore, not override. PTAL

**Comment by seungduk-yanolja** [comment]
@goerch should I abandon this PR or you will review it?

**Review by v-kamelowy — APPROVED** [review]
Can confirm it is working.

`Loading model file Mistral-7B-OpenOrca\pytorch_model-00001-of-00002.bin
Loading model file Mistral-7B-OpenOrca\pytorch_model-00001-of-00002.bin
Loading model file Mistral-7B-OpenOrca\pytorch_model-00002-of-00002.bin
params = Params(n_vocab=32002, n_embd=4096, n_layer=32, n_ctx=32768, n_ff=14336, n_head=32, n_head_kv=8, f_norm_eps=1e-05, f_rope_freq_base=10000.0, f_rope_scale=None, ftype=<GGMLFileType.MostlyF16: 1>, path_model=WindowsPath('Mistral-7B-OpenOrca'))
[...]
[291/291] Writing tensor output.weight                          | size  32002 x   4096  | type F16  | T+  21
Wrote Mistral-7B-OpenOrca_f16.gguf`

**Comment by TheBloke** [comment]
Thanks for this!

File: convert.py [hunk 1]
```diff
@@ -366,44 +366,47 @@ def __init__(self, fname_tokenizer: Path, fname_added_tokens: Path | None) -> No
...
 
-        items = sorted(added_tokens.items(), key=lambda text_idx: text_idx[1])
-        self.added_tokens_list = [text for (text, idx) in items]
-        self.vocab_size_base: int = vocab_size
-        self.vocab_size: int = self.vocab_size_base + len(self.added_tokens_list)
-        self.fname_tokenizer = fname_tokenizer
-        self.fname_added_tokens = fname_added_tokens
+        new_tokens: dict[int, str] = {id: piece for piece, id in added_tokens.items() if id >= vocab_size}
+        expected_new_ids: list[int] = list(range(vocab_size, vocab_size + len(new_tokens)))
+        actual_new_ids: list[int]   = sorted(new_tokens.keys())
```
Comment (cebtenzzre) [inline convert.py:372]: Are these type annotations actually necessary? Usually type checkers like mypy can infer the type of a variable based on the value that is assigned to it.

File: convert.py [hunk 2]
```diff
@@ -366,44 +366,47 @@ def __init__(self, fname_tokenizer: Path, fname_added_tokens: Path | None) -> No
...
+
+        if expected_new_ids != actual_new_ids:
+            raise Exception(f"Expected new token IDs {expected_new_ids} to be sequential; got {actual_new_ids}")
```
Comment (cebtenzzre) [inline convert.py:375]: Do not throw base `Exception`, instead prefer specific exceptions like `ValueError`.

**Review by ggerganov — APPROVED** [review]
@seungduk-yanolja We should merge after addressing @cebtenzzre's review

**Comment by KerfuffleV2** [comment]
This is what the diff looks like with the merge conflict resolved and @cebtenzzre's suggestions applied to c8d6a1f34ab6f1b6bd468d256e535a61f98f114c:

<details>

<summary>Expand</summary>

```patch
diff --git a/convert.py b/convert.py
index 0680f71..655159d 100755
--- a/convert.py
+++ b/convert.py
@@ -366,38 +366,41 @@ class SentencePieceVocab:
             added_tokens = {}
 
         vocab_size: int = self.sentencepiece_tokenizer.vocab_size()
-        expected_ids = list(range(vocab_size, vocab_size + len(added_tokens)))
-        actual_ids   = sorted(added_tokens.values())
-        if expected_ids != actual_ids:
-            raise Exception(f"Expected added token IDs to be sequential and start at {vocab_size}; got {actual_ids}")
 
-        items = sorted(added_tokens.items(), key=lambda text_idx: text_idx[1])
-        self.added_tokens_list = [text for (text, idx) in items]
-        self.vocab_size_base: int = vocab_size
-        self.vocab_size: int = self.vocab_size_base + len(self.added_tokens_list)
-        self.fname_tokenizer = fname_tokenizer
+        new_tokens       = {id: piece for piece, id in added_tokens.items() if id >= vocab_size}
+        expected_new_ids = list(range(vocab_size, vocab_size + len(new_tokens)))
+        actual_new_ids   = sorted(new_tokens.keys())
+
+        if expected_new_ids != actual_new_ids:
+            raise ValueError(f"Expected new token IDs {expected_new_ids} to be sequential; got {actual_new_ids}")
+
+        # Token pieces that were added to the base vocabulary.
+        self.added_tokens_list  = [new_tokens[id] for id in actual_new_ids]
+        self.vocab_size_base    = vocab_size
+        self.vocab_size         = self.vocab_size_base + len(self.added_tokens_list)
+        self.fname_tokenizer    = fname_tokenizer
         self.fname_added_tokens = fname_added_tokens
 
     def sentencepiece_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
         tokenizer = self.sentencepiece_tokenizer
-        for i in range(tokenizer.vocab_size()):
-            piece = tokenizer.id_to_piece(i)
+        for id in range(tokenizer.vocab_size()):
+            piece = tokenizer.id_to_piece(id)
             text: bytes = piece.encode("utf-8")
-            score: float = tokenizer.get_score(i)
+            score: float = tokenizer.get_score(id)
 
             toktype = gguf.TokenType.NORMAL
-            if tokenizer.is_unknown(i):
+            if tokenizer.is_unknown(id):
                 toktype = gguf.TokenType.UNKNOWN
-            if tokenizer.is_control(i):
+            if tokenizer.is_control(id):
                 toktype = gguf.TokenType.CONTROL
 
             # NOTE: I think added_tokens are user defined.
             # ref: https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto
             # if tokenizer.is_user_defined(i): toktype = gguf.TokenType.USER_DEFINED
 
-            if tokenizer.is_unused(i):
+            if tokenizer.is_unused(id):
                 toktype = gguf.TokenType.UNUSED
-            if tokenizer.is_byte(i):
+            if tokenizer.is_byte(id):
                 toktype = gguf.TokenType.BYTE
 
             yield text, score, toktype
```

</details>

I changed it to just use `self.added_tokens_list` instead of `self.new_tokens_list` because otherwise the `__repr__` implementation is broken and `__init__` higher up creates the `added_tokens_list` field. There's no particular reason to change the name here.

Seems like the original author of this pull hasn't been active for a bit.

**Comment by ggerganov** [comment]
Made a new PR with the proposed patch: https://github.com/ggerganov/llama.cpp/pull/3831

**Pull request closed without merging**